File: C++/Arrays/RankTransform.cpp
//optimal
//time complexity:O(nlogn)
//space complexity:O(n)
class Solution {
public:
    vector<int> arrayRankTransform(vector<int>& arr) {
        vector<int> nums = arr;
        sort(nums.begin(), nums.end());
        unordered_map<int, int> mp;
        int n = nums.size();
        int rank = 1;
        for (int i = 0; i < n; i++) {

            if (mp.find(nums[i]) == mp.end()) {
                mp[nums[i]] = rank++;
            }
        }
        vector<int> ans;
        for (int i = 0; i < arr.size(); i++) {
            ans.push_back(mp[arr[i]]);
        }
        return ans;
    }
};

File: Python/Arrays/RankTransform.py
#space complexity-O(n)
#time-Complexity-O(n)
class Solution:
    def arrayRankTransform(self, arr: List[int]) -> List[int]:
        num_rank={}
        sorted_arr=sorted(arr)
        rank=1
        for i in range(len(sorted_arr)):
            if i>0 and sorted_arr[i]>sorted_arr[i-1]:
                rank+=1
            num_rank[sorted_arr[i]]=rank
        for i in range(len(arr)):
            arr[i]=num_rank[arr[i]]
        return arr